[로또] 김홍동

nlom0218 opened the pull request.

nlom0218 committed:
docs: 로또 미션을 위한 기능 목록 작성

nlom0218 committed:
feat: 구입금액 입력 문구 출력

nlom0218 committed:
feat: 구입금액을 입력받은 후 유효성 검사

nlom0218 committed:
feat: 발행한 로또의 개수 출력

구입금액에 1000을 나눈 수가 발행한 로또의 개수이다.

nlom0218 committed:
feat: 발행한 로또의 번호 출력

발행한 로또는 중복되지 않는 1~45 사이의 6개의 숫자이다.
조건을 만족하지 않다면 ERROR로 시작하는 에러 메시지를 출력 후 종료한다.
발행한 로또는 오름차순으로 정렬한 후 배열의 형태의 문자열로 출력한다.

nlom0218 committed:
feat: 당첨 번호 입력 문구 출력

nlom0218 committed:
feat: 당첨 번호를 입력받은 후 유효성 검사

당첨 번호는 쉼표(,)를 기준으로 구분한다.
당첨 번호는 중복되지 않는 1~45 사이의 6개의 숫자이다.
조건을 만족하지 않다면 ERROR로 시작하는 에러 메시지를 출력 후 종료한다.

nlom0218 committed:
refactor: Lotto클래스의 validate 함수 수정

유효성 검사하는 함수를 libs/validation.js에서 관리하도록 한다.

nlom0218 committed:
feat: 보너스 번호를 입력받은 후 유효성 검사

1~45 사이의 한 개의 숫자이어야 한다.
당첨 번호에 포함되어 있지 않아야 한다.
조건을 만족하지 않다면 [ERROR]로 시작하는 에러 메시지를 출력 후 종료한다.

nlom0218 committed:
refactor: 함수명 수정

validation파일의 checkNumbersType과 checkNumbersRange 함수를 각각 isNumberType, isCorrectRange로 수정한다.
조건판별에서 느낌표를 붙여 부정을 나타내기 때문에 is를 접두사로 쓰는 것이 좋아 보인다.

nlom0218 committed:
feat: 당첨 통계 출력

당첨 통계는 당첨 내역과 수익률로 이루어져 있다.

nlom0218 committed:
feat: 당첨 내역 출력

nlom0218 committed:
feat: 수익률 출력

수익률은 소수점 둘째 자리에서 반올림한다.

nlom0218 committed:
feat: 수익률 출력 수정

퍼센트를 구하는 것이기 때문에 로또 상금에 구입한 금액을 나눈 후 100을 나누어야 한다.
100을 나누지 않아 해당 과정을 추가한다.

nlom0218 committed:
feat: 게임을 종료한다.

nlom0218 committed:
docs: 리팩토링 목록 추가

nlom0218 committed:
refactor: WinningNumbers 클래스 분리하기

nlom0218 committed:
refactor: BonusNumber 클래스 분리하기

nlom0218 committed:
reactor: Lottos 클래스 분리하기

기존 Game 클래스를 Lottos 클래스, WinningNumbers 클래스, BonusNumber 클래스로 분리한다.
Lottos 클래스는 입력한 금액에 따라 발행된 로또들을 관리한다. 로또는 Lotto 클래스에서 관리한다.
WinningNumbers 클래스는 당첨 번호, BonusNumber 클래스는 보너스 번호를 관리한다.

nlom0218 committed:
refactor: validation 함수 이동하기

Lottos, WinningNumber, BonusNumber, Lotto 클래스가 생성될 때 생성자 함수에서 validation의
함수 실행하기

nlom0218 committed:
refactor: 유효성 검사 함수 인스턴스 생성 전으로 이동하기

nlom0218 committed:
refactor: 함수(메서드) 이동 및 분리하기

구매 금액에 따른 복수의 로또를 발행하는 함수(Lottos 클래스의 publish 함수)
에서 새로운 Lotto 인서턴스를 생성하는 로직을 분리하기

nlom0218 committed:
docs: 기능 목록 추가

nlom0218 committed:
feat: 에러가 발생하면 에러 메시지를 출력한 후 종료

validation의 함수들은 결과만 반환하는 역할을 한다.
에러를 출력한 후 종료하는 새로운 함수를 만들다.

nlom0218 committed:
refactor: 유효성 검사에 사용되는 객체 이름 수정

validation 객체를 checkValue로 수정한다. 값을 체크한 뒤 errorMsg를 반환한다.
validation 객체 내 함수들의 이름도 간단하게 수정한다.

nlom0218 committed:
refactor: 에러 메시지 상수화, 에러 메시지 생성 함수 분리

에러 메시지를 생성하는 함수를 만들어 분리한다. 에러 메시지 생성 함수는 name 매개변수를 받기 때문에 name인수를 추가한다.
변하지 않는 에러 메시지를 상수화한다.

nlom0218 committed:
refactor: 매직 넘버 상수화하기

구입 금액, 로또 번호, 당첨 번호, 보너스 번호 상수화한다.

nlom0218 committed:
refactor: 유효성 검사 함수 이동하기

구매 금액, 로또 번호, 당첨 번호, 보너스 번호의 유효성 검사는 해당 인스턴스가 생성되는 클래스에서 진행한다.

nlom0218 committed:
refactor: 매직 넘버 상수화하기

플레이어 입력에 해당하는 메시지를 상수화한다.

nlom0218 committed:
refactor: 변수명, 함수명 수정하기

numberList를 numbers로 수정한다.

nlom0218 committed:
feat: 에러를 발생시키기 전 애플리케이션 종료

Console.close() 함수를 사용하여 종료 후 에러 메시지를 출력한다.

nlom0218 committed:
refactor: 변수명, 함수명 수정하기

-Array로 표현된 변수명, 함수명을 -s로 수정한다.
printWinningHistory를 printWinningDetails로 수정한다.
Lotto 클래스의 getResult를 getRank로 수정한다. 또한 해당 함수 내에서 선언한 winninCount를 count로 수정한다.

nlom0218 committed:
docs: 테스트 목록 작성

nlom0218 committed:
test: Lotto 클래스 생성 테스트 작성

로또 클래스를 생성하는 테스트를 예외상황을 기준으로 작성한다.

nlom0218 committed:
test: Lotto 클래스의 getRank 메서드 테스트 작성하기

Lotto 클래스의 getRank 메서드 테스트를 1~5등, 낙첨으로 나누어서 작성한다.

nlom0218 committed:
test: Lottos 클래스 생성 테스트 작성

nlom0218 committed:
test: Lottos 클래스에서 발행된 로또개수 테스트 작성

nlom0218 committed:
test: Lottos 클래스의 로또의 결과 테스트

nlom0218 committed:
test: Lottos 클래스에서 등수 별 당첨 개수 테스트 작성

nlom0218 committed:
test: WinningNumbers 클래스 테스트 작성

WinningNumbers 클래스를 생성하는 테스트를 예외사항을 기준으로 작성한다.

nlom0218 committed:
test: BonusNumber 클래스 테스트 작성

BonusNumber 클래스를 생성하는 테스트를 예외를 기준으로 작성한다.

nlom0218 committed:
test: 유효성 검사 - checkValue.money() 함수 테스트 작성

nlom0218 committed:
test: 유효성 검사 - checkValue.numbers() 함수 테스트 작성

nlom0218 committed:
test: 유효성 검사 - checkValue.bonusNumber() 함수 테스트 작성

nlom0218 committed:
refactor: Lottos 클래스의 printRate 함수를 두 부분을 나누기

함수가 한 가지의 일만 하도록 하기!
기존 수익률을 출력하는 함수를 수익률을 계산하는 함수와 수익률을 출력하는 함수로 나누기

nlom0218 committed:
test: Lottos 클래스의 수익률 계산 함수 테스트 작성

Lottos 클래스에서 수익률 계산하는 함수를 테스트한다. - calculateRate() 테스트

nlom0218 committed:
refactor: 구입 금액, 로또에 관련된 매직 넘버 상수화하기

구입 금액 단위, 최소 구입 금액을 상수화한다.
로또의 최소 숫자, 최대 숫자, 로또 번호의 개수를 상수화한다.

nlom0218 committed:
refactor: 로또의 등수를 가져오는 함수 수정

Lotto 클래스의 getRank() 함수에서 6, 7, 8등도 함께 반환한다.
Lottos 클래스의 getResult() 함수에서 5등 이하의 값만 반환한다.

nlom0218 committed:
refactor: 로또 등수를 가져오는 함수명 및 로또 등수를 나타내는 변수명 수정

Lottos 클래스의 getResult() 함수를 getRanks() 로 이름을 수정한다. 또한 해당 값을 반환하는 변수도 lottoRanks로 수정한다.
Lotto 클래스에서는 getRank() 함수로 등수를 받아온다. 이에 통일감을 주기 위해 수정한다.

nlom0218 committed:
test: 로또 등수를 가져오는 로직이 변경함에 따른 테스트 코드 수정

로또 등수를 가져오는 로직이 변경되어 이에 해당하는 테스트 코드를 수정한다.
변수명 또한 통일감을 주기 위해 수정한다.

nlom0218 committed:
refactor: 매직 넘버 상수화하기

당첨 세부 사항 문자열, 상금 숫자, 1등, 2등, 마지막 등수를 상수화한다.
추가적으로 이전에 미쳐 변경하지 못하였던 변수명도 함께 수정한다.(lottoRanks)

nlom0218 committed:
docs: 디렉토리 구조 추가

nlom0218 committed:
refactor: 미션 제출 후 리팩터링 목록 추가

nlom0218 committed:
refactor: InputView 객체는 사용자의 입력을 위한 출력을 담당

nlom0218 committed:
refactor: OutputView 객체 분리하기

OutputView 객체는 메시지 출력을 담당한다.
OutputView의 메서드들은 게임의 흐름을 담당하는 곳에서 사용한다.

nlom0218 committed:
refactor: Lottos 클래스명 PurchasedLottos로 변경하기

nlom0218 committed:
refactor: PurchasedLottos의 count 필드 제거하기

nlom0218 committed:
refactor: 유효성 검사 실시를 위한 Validator 객체 생성

유효성 검사는 Model 클래스에서 하지 않고 컨트롤러에서 진행한다.
기존 checkValue 객체는 삭제한다.

nlom0218 committed:
refactor: Validator 객체를 libs 폴더로 옮기기

nlom0218 committed:
refactor: 종료를 위한 Quit 객체 생성하기

nlom0218 committed:
refactor: Validator 메서드 분리하기

nlom0218 committed:
refactor: winningPhrases, lottoPrizes 클래스 필드로 변경하기

nlom0218 committed:
refactor: ERROR_HADEING을 만들어 분리하기

nlom0218 committed:
feat: 수익률 자리수 추가하기

nlom0218 committed:
refactor: PurchasedLottos를 PlayerLottos로 변경 및 콜백 분리하기

nlom0218 committed:
refactor: IOView객체 생성하여 View 관리하기

nlom0218 committed:
refactor: OutputView에서 결과 및 수익률 계산하기

nlom0218 committed:
refactor: 모델 클래스의 클래스 필드 선언

nlom0218 committed:
refactor: Lotto의 getRank 메서드 수정

Sallyvely commented on src/View/OutputView.js:
```diff
@@ -0,0 +1,101 @@
+  },
+
+  printLottosCount(count) {
+    Console.print(`\n${count}개를 구매했습니다.`);
```
구매했다는 문구도  상수로 추가하면 좋을 것 같습니다 !

nlom0218 replied:
네! 감사합니다ㅎㅎ

Sallyvely commented on src/App.js:
```diff
@@ -1,5 +1,83 @@
+    this.#playerLottos.publish(money);
+
+    OutputView.printLottosCount(this.#playerLottos.getCount());
+    OutputView.printLottos(this.#playerLottos.getLottos());
```
30,31 라인을 새로운 메서드로 합쳐서 작성하면 가독성이 더 좋을 것 같습니다. 개인적인 생각이니 참고만 해주세용 ☺️

nlom0218 replied:
메서드를 하나 더 만들어서 불러오는 것을 말씀하시는거죠!?

sangpok replied:
> 30,31 라인을 새로운 메서드로 합쳐서 작성하면 가독성이 더 좋을 것 같습니다. 개인적인 생각이니 참고만 해주세용 ☺️

저도 같은 생각입니다! 같은 playerLottos에 대한 정보 출력이니 하나의 메서드로 호출하게끔 해도 괜찮을 것 같습니다 :)

nlom0218 replied:
네! 말씀해주신 내용들이 훨씬 깔끔하고 기능 분리가 잘 이루어질 수 있을 듯 합니다ㅎㅎ

Sallyvely commented:
상수로 정리해두신게 정말 깔끔하네요. 많이 배우고 갑니다🤓

sangpok commented on src/App.js:
```diff
@@ -1,5 +1,83 @@
+  }
+
+  handleReadMoney(money) {
+    money = Number(money);
+
+    Validator.money(money);
```
저는 유효성 검사를 하고 나서 형변환을 해주는 순서로 코드를 짰었는데요, 그 이유는 올바른 값인지 먼저 체크하고 나서 형변환 및 저장을 하는 게 적절한 순서라고 생각했었습니다. 

그런데 반대로 홍동님은 형변환을 하고 나서 유효성 검사를 진행하시는데, 어차피 number형으로 변환을 해야 검사가 되니까 그렇게 하신걸까요? 저랑 달라서 여쭤봅니다ㅎㅎ

nlom0218 replied:
현재 상황에서 숫자형으로 먼저 바꾼 이유는 `Validator` 객체의 `checkNumberType` 메서드 때문이었던 거 같아요. `Number.isNaN()` 메서드로 숫자인지 아닌지를 파악을 하고 있어 미리 변경을 했어요. 즉, 문자열에 숫자가 아닌 값이 포함되었을 때 `Number.isNaN()`의 값이 `true`로 나오게 했습니다.

이 문제는 유효성 검사를 실시할 때 `Number.isNaN()`를 사용하는 것이 아니라 정규식을 사용하면 해결될 듯 해요!

추가적으로 언제 형변화를 해야 하는가?에 대해서는 언뜻 기억나는 것이 있어요. '형변환이 필요한 경우 상단에서 진행한다'라는 것을 어디서 봤는데... 기억이 가물가물하네요ㅠㅠㅎㅎ

추가 리팩터링을 통해 유효성 검사를 정규식으로 바꾸고 저장하는 값도 생각을 해봐야겠습니다!

sangpok commented on src/App.js:
```diff
@@ -1,5 +1,83 @@
+  }
+
+  handleWinningNumbers(winningNumbers) {
+    winningNumbers = winningNumbers.split(',').map((item) => Number(item));
```
구분자 `,` 도 상수화로 처리하면 좋을 것 같습니다! 나중에 구분자가 `/`로 바뀔 수 있다는 가정을 해본다면요ㅎㅎ

sangpok commented on src/App.js:
```diff
@@ -1,5 +1,83 @@
+  handleWinningNumbers(winningNumbers) {
+    winningNumbers = winningNumbers.split(',').map((item) => Number(item));
+
+    Validator.numbers(winningNumbers, WINNING_NUMBER);
```
문득 생각이 난 건데, 홍동님은 이 코드 뿐만 아니라 `[기능].[대상]` 형태의 네이밍이 익숙하신 것 같아요! 이것도 코드 컨벤션인지 궁금합니다. 저는 호출하는 메서드 이름에 조금 더 집중하는 편이라, `numbers()`를 호출하면 어떤 기능을 하는 거지? 라는 생각을 했었습니다ㅎㅎ

nlom0218 replied:
저도 우선 호출하는 메서드의 이름에 집중을 하는 편이에요. 이후 떠오른 메서드 이름에 객체의 이름이 포함되어 있으면 해당 부분은 생략해서 메서드 이름을 짓고 있어요. 예를 들어 위의 `Validator.numbers`를 `Validator.validationNumbers`로 지었다면 `validation`이라는 부분은 객체의 이름에서 충분히 유추할 수 있을 테니까요!

추가적으로 상수를 만드는 부분에서도 많이 적용을 했습니다! MESSAGE라는 객체를 만들고 어떤 메시지인지만 간략하게 `key`를 지정했습니다!

sangpok commented on src/App.js:
```diff
@@ -1,5 +1,83 @@
+  handleReadBonusNumber(bonusNumber) {
+    bonusNumber = Number(bonusNumber);
+
+    Validator.bonusNumber(bonusNumber, this.#winningNumbers);
```
보너스 번호를 검사할 때 validator에서 winningnumbers를 가져오는 게 아니라 넘겨주는 형태네요! 책임 분리를 잘 하신 것 같습니다👍👍

sangpok commented on src/Model/Lotto.js:
```diff
@@ -0,0 +1,37 @@
+  getNumbers() {
+    this.#numbers.sort((a, b) => a - b);
+
+    return `[${this.#numbers.join(', ')}]`;
```
여기도 만약 출력하는 형태가 `[1, 2, 3, 4, 5, 6]` 형태가 아니라 `(1, 2, 3, ~ )` 같은 형태로 바뀔 수도 있다면 상수화가 필요한 부분이라고 생각이 듭니다😊🙌

nlom0218 replied:
오호! 이런 부분도 상수화가 필요하겠네요!

sangpok commented on src/Model/Lotto.js:
```diff
@@ -0,0 +1,37 @@
+  getRank(winningNumbers, bonusNumber) {
+    let count = 0;
+
+    this.#numbers.forEach((number) => {
+      if (winningNumbers.includes(number)) count += 1;
+    });
```
`forEach()`는 O(n), `includes()`도 O(n)의 시간 복잡도를 가지고 있어서 합치면 O(n^2)의 시간 복잡도가 걸리게 되므로 다른 방법을 생각해보는 것도 도움이 될 것 같습니다! (물론 데이터 양이 작다면 차이는 미미하지만요ㅎㅎ)

다른 분들은 `Set`을 이용해서 포함되어있는지 아닌지 체크하는 방법을 많이 쓰는 것 같습니다.

nlom0218 replied:
좋은 팁 감사합니다! `Set` 객체를 유용하게 사용해봐야겠어요ㅎㅎ

sangpok commented on src/Model/Lotto.js:
```diff
@@ -0,0 +1,37 @@
+
+    if (count === 6) return PLACE.FIRST;
+
+    if (count === 5 && this.#numbers.includes(bonusNumber)) return PLACE.SECOND;
```
조건 분기문에선 직접 메소드를 호출하는 것보단 좀 더 명시적으로 흐름을 읽을 수 있는 boolean 변수를 활용하시는 것을 추천드립니다!

```javascript
const hasBonusNumber = this.#numbers.includes(bonusNumber);
if (count === 5 && hasBonusNumber) return PLACE.SECOND;
```

nlom0218 replied:
그러쿤요! 하나 더 배워갑니다!

sangpok commented on src/Model/PlayerLottos.js:
```diff
@@ -0,0 +1,56 @@
+  }
+
+  publish(money) {
+    for (let num = 0; num < money / MONEY.UNIT; num++) {
+      const newLotto = this.createNewLotto();
+      this.#list.push(newLotto);
+    }
```
인덱스가 필요없는 상황이라면 `while(조건문)`을 이용해보시는 건 어떨까요? 👍

```javascript
const lottoCount = money / MONEY.UNIT
while (this.#list.length < lottoCount) {
  const newLotto = this.createNewLotto();
  this.#list.push(newLotto)
}
```

nlom0218 replied:
이런 방법도 있네요! 감사합니다~ 참고하겠습니다!ㅎㅎ

sangpok commented on src/Model/PlayerLottos.js:
```diff
@@ -0,0 +1,56 @@
+    return this.#list.length;
+  }
+
+  getLottos() {
+    const lottos = [];
+
+    this.#list.forEach((lotto) => {
+      const lottoNumbers = lotto.getNumbers();
+
+      lottos.push(lottoNumbers);
+    });
+
+    return lottos;
+  }
```
`getLottos()` 메소드는 `map()` 메소드를 활용하면 깔끔해질 것 같습니다 !

```javascript
getLottos() {
  return this.#list.map(item => item.getNumbers());
}
```

nlom0218 replied:
그러네요! 생각지도 못했어요ㅋㅋㅋ😂

sangpok commented on src/Model/PlayerLottos.js:
```diff
@@ -0,0 +1,56 @@
+    });
+
+    return lottoRanks.filter((rank) => rank <= PLACE.LAST);
+  }
```
여기도 `map()` 메소드를 활용하면 좋을 것 같습니다 :)

```javascript
getRanks(winningNumbers, bonusNumber) {
  return this.#list
    .map((item) => item.getRank(winningNumbers, bonusNumber))
    .filter((rank) => rank <= PLACE.LAST);
}
```

sangpok commented on src/View/IOView.js:
```diff
@@ -0,0 +1,23 @@
+      Console.close();
+    },
+
+    withErrorMessage(errorMsg) {
+      Console.close();
+
+      throw new Error(`${ERROR_HEADING} ${errorMsg}`);
+    },
```
IOView라는 입출력을 담당하는 핸들러를 만드셨군요 ! :) Quit이라는 객체를 이렇게 살릴 수도 있겠네요👍

다만 생각이 드는 부분은 `withErrorMessage()` 라는 메소드인데, 저기에서 `new Error()` 객체를 생성한다면 저길 기점으로 에러 스택이 생성이 되어서 나중에 에러 트레이싱이 힘들어진다고 합니다. 때문에 다른 방식으로 접근해야 하지 않을까 싶습니다😊😊

nlom0218 replied:
그러면 `throw new Error()` 부분도 다른 객체를 만들어 관리하는 것이 좋겠네요!

sangpok commented on src/View/OutputView.js:
```diff
@@ -0,0 +1,101 @@
+  },
+
+  getRate(lottosRank, lottoCount) {
+    const finalPrize = this.lottoPrizes.reduce((acc, cur, idx) => {
+      const winningCount = this.calculateWinningCount(lottosRank, idx);
+
+      return acc + cur * winningCount;
+    }, 0);
```
앗 이렇게 최종 합산 금액을 구하는 방법 좋네요👍👍 좀 더 명시적으로 함수를 빼서 finalPrize는 호출하면 담길 수만 있도록 하는 로직도 괜찮을 것 같습니다!

```javascript
#getFinalPrize() {
  return this.lottoPrizes.reduce((acc, cur, idx) => {
      const winningCount = this.calculateWinningCount(lottosRank, idx);

      return acc + cur * winningCount;
    }, 0);
}

const finalPrize = this.#getFinalPrize();
```

nlom0218 replied:
최종금액을 구하는 메서드를 하나 더 만드는 것이네요! 주현님 말씀처럼 역할 분리를 하는 것이 좋아보입니다!

sangpok commented on src/View/OutputView.js:
```diff
@@ -0,0 +1,101 @@
+    );
+  },
+
+  setNumberDigits(rate) {
+    const integer = rate.slice(0, -2).split('').reverse();
+    const decimal = rate.slice(-2);
+
+    let newInteger = [];
+    integer.forEach((num, idx) => {
+      newInteger.push(num);
+      if (this.isPutComma(integer, idx)) newInteger.push(',');
+    });
+
+    return newInteger.reverse().join('') + decimal;
+  },
```
`toLocaleString()` 라는 메소드 한번 찾아보시면 억울하실 것 같은 이 느낌...😂😂

nlom0218 replied:
핫핫핫... 억울하진 않지만,, 웃음이 나옵니다😂😂

sangpok commented on src/libs/Validator.js:
```diff
@@ -0,0 +1,111 @@
+
+  checkNumberType(money) {
+    if (Number.isNaN(money))
+      return Quit.withErrorMessage(createErrorMsg.type(PURCHASE_MONEY));
```
`createErrorMsg` 객체 아이디어 좋은 것 같습니다 👍👍

sangpok commented on src/libs/Validator.js:
```diff
@@ -0,0 +1,111 @@
+  },
+};
+
+const createErrorMsg = {
+  range: (name) => {
+    return `${name}: ${LOTTO.MIN_NUMBER}~${LOTTO.MAX_NUMBER} 사이의 값만 입력할 수 있습니다.`;
+  },
+
+  type: (name) => {
+    return `${name}: 숫자만 입력할 수 있습니다.`;
+  },
+
+  length: (name) => {
+    return `${name}: ${LOTTO.NUMBERS_COUNT}개의 숫자로 이루어져야 합니다.`;
+  },
+
+  overlap: (name) => {
+    return `${name}: 중복되지 않는 숫자로 이루어져야 합니다.`;
+  },
+};
```
객체에서 메소드를 축약형인 `range(name) { }`이 아닌 Arrow Function으로 담으신 이유가 궁금합니다!

호출하기 전까지 생성이 필요 없는 경우라면 arrow function으로 지정한다 뭐 그런 말이 있었던 것 같은데 정확하게는 몰라서😂

nlom0218 replied:
메서드에서 `arrow function` 사용에 대해 찾아봤는데 `화살표 함수로 메소드를 정의하는 것은 피해야 한다.`라고 합니다! 그 이유는 `this`를 사용할 경우 메서드를 호출한 객체가 아닌 `window`를 가리키게 된다고 합니다. 제가 이건 리팩터링을 해야겠어요ㅎㅎ

https://poiemaweb.com/es6-arrow-function 의 4번째 파트인 `화살표 함수를 사용해서는 안되는 경우`를 참고했습니다!

sangpok commented on src/libs/const.js:
```diff
@@ -0,0 +1,67 @@
+  FIFTH: '3개 일치 (5,000원)',
+};
+
+const PRIZE = {
+  FIRST: 2000000000,
+  SECOND: 30000000,
+  THIRD: 1500000,
+  FOURTH: 50000,
+  FIFTH: 5000,
+};
```
자바스크립트에서는 가독성을 위해 숫자를 표현할 때 언더바 `_`로 자릿수를 구분 지을 수 있다는 것 참고하시면 좋을 것 같습니다👍

```suggestion
const PRIZE = {
  FIRST: 2_000_000_000,
  SECOND: 30_000_000,
  THIRD: 1_500_000,
  FOURTH: 50_000,
  FIFTH: 5_000,
};
```

nlom0218 replied:
좋은 꿀팁감사합니다!

sangpok reviewed (commented):
역할 분리에 대해서 노력하신 점들이 보여서 좋은 코드라고 생각했습니다👍 수고하셨습니다 !!

nlom0218 committed:
refactor: 숫자 타입인지 확인하는 유효성 검사 수정하기

isNaN이 아닌 정규식으로 판별한다.

nlom0218 committed:
refactor: 플레이어의 로또정보 출력 메서드 분리

nlom0218 committed:
refactor: 구분자 상수화하기 및 숫자형 변환 위치 변경

nlom0218 committed:
refactor: Lotto getNumbers에서 문자열 상수화하기

nlom0218 committed:
refactor: Lotto.getRank 메서드에서 Set 객체 활용

forEach 메서드내부에서 includes메서드를 사용하고 있기 때문에 시간 복잡도가 O(n^2)이다.
이를 줄이기 위해 winningNumbers 배열을 Set 객체로 바꾸어 has 메서드를 사용한다.

nlom0218 committed:
refactor: Lotto.getRank 메서드 분리

순위를 반환하는 메서드와 순위를 계산하는 메서드 그리고 보너스 번호가 있는지 확인하는 메서드로 분리한다.

nlom0218 committed:
refactor: PlayerLottos의 getLottos, getRanks 메서드에서의 map메서드 사용

forEach가 아니라 map메서드를 사용하여 바로 반환한다.

nlom0218 committed:
refactor: errorHandelr 함수 분리하기

nlom0218 committed:
refactor: calculateFinalPrize 함수 분리

nlom0218 committed:
refactor: 메서드 함수선언으로 수정하기

nlom0218 committed:
refactor: const PRIZE 언더바 추가하기
